Terraform support for Nexus (#3550)
temporalio/documentation: docs/production-deployment/cloud/terraform-provider.mdx

The Namespace is now a part of the Terraform state and all changes to the Namespace should be managed by Terraform.
Once a resource has been imported into Terraform, outside changes to the resource will create Terraform "drift" errors on subsequent Terraform operations.

## Manage Temporal Cloud Nexus Endpoints with Terraform

Terraform is a great way to automate the management of [Nexus Endpoints](/nexus/endpoints).
The provider allows you to import, create, update, and delete Nexus Endpoints with Terraform.

You must use an Identity with Developer role (or higher) and Namespace Admin permissions on the Endpoint's target Namespace.

> This includes the Account Owner, Global Admin, or Developer Account Role.

Reviewer: [nit] The second sentence addresses the first part of the initial statement (Account level role) but is not relevant to the second part (namespace level permissions), making the start with "this" somewhat awkward.

     regions        = ["aws-us-east-1"]
     api_key_auth   = true
     retention_days = 14
     timeouts {
       create = "10m"
       delete = "10m"
     }
   }

   resource "temporalcloud_namespace" "caller_namespace_2" {
     name           = "terraform-caller-namespace-2"
     regions        = ["gcp-us-central1"]
     api_key_auth   = true
     retention_days = 14
     timeouts {
       create = "10m"
       delete = "10m"
     }
   }

   resource "temporalcloud_nexus_endpoint" "nexus_endpoint" {
     name        = "terraform-nexus-endpoint"
     description = <<-EOT
       Service Name:
         my-hello-service
       Operation Names:
         echo
         say-hello

       Input / Output arguments are in the following repository:
       https://github.com/temporalio/samples-go/blob/main/nexus/service/api.go
     EOT
     worker_target = {
       namespace_id = temporalcloud_namespace.target_namespace.id
       task_queue   = "terraform-task-queue"
     }
     allowed_caller_namespaces = [
       temporalcloud_namespace.caller_namespace.id,
       temporalcloud_namespace.caller_namespace_2.id,
     ]
   }
   ```

   In this example, 3 Namespaces are created:

>    - target Namespace for a Nexus Endpoint - Nexus requests will be routed to a Worker in the target Namespace.

Reviewer: [nit] Nexus requests will be routed to a Worker that polls the target Namespace.

The Terraform Provider docs show how to generate CA certs within Terraform configuration files and create a Namespace with API Key based authentication.

**How do I validate the creation of the Nexus Endpoint?**

You can validate the creation of the Nexus Endpoint through the Temporal Web UI or through the `tcld nexus endpoint get` command.

**Using the Temporal Web UI**

1. Log into the Temporal Cloud Web UI.
1. Navigate to [the Nexus page](https://cloud.temporal.io/nexus).
1. Search for the Nexus Endpoint you created, using only the Nexus Endpoint Name (without an account suffix).

**Using the tcld CLI utility**

Validate the creation of your Nexus Endpoint through the Terraform provider.
To validate see your Nexus Endpoint in the Cloud UI or through the `tcld nexus endpoint get` command.

Run the command below with your Nexus Endpoint Name (without an account suffix):

```bash
tcld nexus endpoint get -n "<your-nexus-endpoint>"
```

**How do I update a Nexus Endpoint?**

Terraform automatically recognizes changes made within `.tf` files and applies those changes to Temporal.

> For example, change the retention period setting in the Terraform file from the previous example and watch Terraform apply the change without any additional steps required by you.

Reviewer: Perhaps changing this to the `task-queue` property is a better example since it is on the nexus endpoint (versus the retention period which is a namespace property).

Reviewer: Or maybe the intention was `allowed_caller_namespaces ` based on the example below?

Author: thanks for catching this!

1. Add or remove allowed caller Namespaces by updating the Nexus Endpoint configuration to remove `caller_namespace_2`:

   ```yml
   resource "temporalcloud_nexus_endpoint" "nexus_endpoint" {
     name        = "terraform-nexus-endpoint"
     description = <<-EOT
       Service Name:
         my-hello-service
       Operation Names:
         echo
         say-hello

       Input / Output arguments are in the following repository:
       https://github.com/temporalio/samples-go/blob/main/nexus/service/api.go
     EOT
     worker_target = {
       namespace_id = temporalcloud_namespace.target_namespace.id
       task_queue   = "terraform-task-queue"
     }
     allowed_caller_namespaces = [
       temporalcloud_namespace.caller_namespace.id
     ]
   }
   ```

> 1. Apply your configuration.

Reviewer: [nit] step 2.

Reviewer: Same problem elsewhere I am not sure if it is intentional.

Author: yes intentional, based on the rest of this docs page using repeated 1. thanks for flagging though.

   When prompted, answer yes to continue:

   ```command
   terraform apply
   ```

   Upon completion, you will see a success message indicating your Nexus Endpoint has been updated.

>    It may take several seconds to update a Nexus Endpoint.

Reviewer: [nit] The customers can verify the update from the UI and with the tcld command within several seconds, but the propagation to the data plane & change detection by the workers might take up to a minute.

Author: fixed

   ```

**How do I delete a Nexus Endpoint?**

To delete the Nexus Endpoint, run the following command:

```bash
terraform destroy
```

Upon completion, you will see a success message indicating all resources have been deleted.

```text
Destroy complete! Resources: 4 destroyed.
```

**How do I import a Temporal Cloud Nexus Endpoint?**

If you have an existing Nexus Endpoint in Temporal Cloud, you can import it into Terraform to manage the Nexus Endpoint from Terraform using the `terraform import` command.

1. Initialize the Terraform provider in a new directory.

   Run the following command to initialize the Terraform provider.

   ```bash
   terraform init
   ```

1. Provide a configuration placeholder in your Terraform configuration and ensure your [API key is provided](#setup).

   ```yml
   terraform {
     required_providers {
       temporalcloud = {
         source = "temporalio/temporalcloud"
       }
     }
   }

   provider "temporalcloud" {
   }

   resource "temporalcloud_nexus_endpoint" "nexus_endpoint" {
   }
   ```

1. Run the `terraform import` command from the command line and pass in the Nexus Endpoint ID.

>    Your Nexus Endpoint ID is available at the top of the Nexus Endpoint's page in the [Temporal Cloud UI](https://cloud.temporal.io/nexus).

Reviewer: [nit] a screenshot can be helpful

Author: great idea, added!

     description = <<-EOT
       Service Name:
         my-hello-service
       Operation Names:
         echo
         say-hello

       Input / Output arguments are in the following repository:
       https://github.com/temporalio/samples-go/blob/main/nexus/service/api.go
     EOT
     worker_target = {
       namespace_id = temporalcloud_namespace.target_namespace.id
       task_queue   = "terraform-task-queue"
     }
     allowed_caller_namespaces = [
       temporalcloud_namespace.caller_namespace.id,
       temporalcloud_namespace.caller_namespace_2.id,
     ]
   }
   ```

   In this example, 3 Namespaces are created:
   - target Namespace for a Nexus Endpoint - Nexus requests will be routed to a Worker that polls the target Namespace.
   - caller Namespace(s) - Nexus Operations are invoked from caller Namespace, for example from a caller Workflow.

   These Namespaces are referenced in the [Nexus Endpoint](/nexus/endpoints) configuration:
   - `worker_target` (Namespace and Task Queue) - currently only a single worker_target is supported.
   - `allowed_caller_namespaces` - used to enforce Nexus Endpoint [runtime access controls](/nexus/security#runtime-access-controls).

> 

Reviewer 2: ```suggestion
```

   ```bash
   temporalcloud_nexus_endpoint.nexus_endpoint: Creation complete after 2s [id=b158063be978471fa1d200569b03834d]
   ```

You can find more examples of Nexus Endpoint management in the Terraform Provider docs located on HashiCorp's [Terraform Registry](https://registry.terraform.io/providers/temporalio/temporalcloud/latest/docs/resources/nexus_endpoint).
The Terraform Provider docs show how to generate CA certs within Terraform configuration files and create a Namespace with API Key based authentication.

**How do I validate the creation of the Nexus Endpoint?**

You can validate the creation of the Nexus Endpoint through the Temporal Web UI or through the `tcld nexus endpoint get` command.

**Using the Temporal Web UI**

1. Log into the Temporal Cloud Web UI.
1. Navigate to [the Nexus page](https://cloud.temporal.io/nexus).
1. Search for the Nexus Endpoint you created, using only the Nexus Endpoint Name (without an account suffix).

**Using the tcld CLI utility**

Validate the creation of your Nexus Endpoint through the Terraform provider.
To validate see your Nexus Endpoint in the Cloud UI or through the `tcld nexus endpoint get` command.

> Run the command below with your Nexus Endpoint Name (without an account suffix):

Reviewer 2: ```suggestion
Run the below command using your Nexus Endpoint Name.
Do not use the account ID suffix with this endpoint Name:
```

The Namespace is now a part of the Terraform state and all changes to the Namespace should be managed by Terraform.
Once a resource has been imported into Terraform, outside changes to the resource will create Terraform "drift" errors on subsequent Terraform operations.

## Manage Temporal Cloud Nexus Endpoints with Terraform

> Terraform is a great way to automate the management of [Nexus Endpoints](/nexus/endpoints).

Reviewer 2: ```suggestion
Terraform provides a great way to automate the management of [Nexus Endpoints](/nexus/endpoints).
```

The provider allows you to import, create, update, and delete Nexus Endpoints with Terraform.

You must use an Identity with [Developer role (or higher)](/cloud/users#account-level-roles) and [Namespace Admin permission](/cloud/users#namespace-level-permissions) on the Endpoint's target Namespace.

**How do I create a Nexus Endpoint with Terraform?**

> 

Reviewer 2: ```suggestion
```

   ```

You can find more examples of Nexus Endpoint management in the Terraform Provider docs located on HashiCorp's [Terraform Registry](https://registry.terraform.io/providers/temporalio/temporalcloud/latest/docs/resources/nexus_endpoint).
The Terraform Provider docs show how to generate CA certs within Terraform configuration files and create a Namespace with API Key based authentication.

**How do I validate the creation of the Nexus Endpoint?**

You can validate the creation of the Nexus Endpoint through the Temporal Web UI or through the `tcld nexus endpoint get` command.

**Using the Temporal Web UI**

1. Log into the Temporal Cloud Web UI.
1. Navigate to [the Nexus page](https://cloud.temporal.io/nexus).
1. Search for the Nexus Endpoint you created, using only the Nexus Endpoint Name (without an account suffix).

**Using the tcld CLI utility**

Validate the creation of your Nexus Endpoint through the Terraform provider.
To validate see your Nexus Endpoint in the Cloud UI or through the `tcld nexus endpoint get` command.

Run the below command using your Nexus Endpoint Name.

> Do not use the account ID suffix with this endpoint Name:

Reviewer 2: ```suggestion
Do not use the account ID suffix with this endpoint name:
```

```bash

> tcld nexus endpoint get -n "<your-nexus-endpoint>"

Reviewer 2: ```suggestion
tcld nexus endpoint get -n "<your-nexus-endpoint-name-without-account-suffix>"
```

```

**How do I update a Nexus Endpoint?**

Terraform automatically recognizes changes made within `.tf` files and applies those changes to Temporal.

> For example, change the allowed caller Namespaces on a Nexus Endpoint.

Reviewer 2: ```suggestion
For example, to change the allowed caller Namespaces on a Nexus Endpoint:
```
